universal-application-tool-0.0.1/app/services/cloud/azure/BlobStorage.java

```diff
@@ -0,0 +1,241 @@
+import com.azure.storage.blob.models.BlobServiceProperties;
+import com.azure.storage.blob.models.UserDelegationKey;
+import com.azure.storage.blob.sas.BlobSasPermission;
+import com.azure.storage.blob.sas.BlobServiceSasSignatureValues;
+import com.azure.storage.common.sas.SasProtocol;
+import com.typesafe.config.Config;
+import java.net.URL;
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.util.List;
+import java.util.UUID;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import play.Environment;
+import services.cloud.StorageClient;
+import services.cloud.StorageServiceName;
+
+/**
+ * SimpleStorage provides methods to create federated links for users of CiviForm to upload and
+ * download files directly to and from AWS Simple Storage Service (S3).
+ */
+@Singleton
+public class BlobStorage implements StorageClient {
+
+  public static final String AZURE_STORAGE_ACCT_CONF_PATH = "azure.blob.account";
+  public static final String AZURE_CONTAINER_CONF_PATH = "azure.blob.container";
+  public static final String AZURE_REGION_CONF_PATH = "java.time.zoneid";
+  public static final Duration AZURE_SAS_TOKEN_DURATION = Duration.ofMinutes(10);
+  public static final Duration AZURE_USER_DELEGATION_KEY_DURATION = Duration.ofMinutes(60);
+
+  private final Credentials credentials;
+  private final String container;
+  private final Client client;
+  private final String accountName;
+  private final String blobEndpoint;
+  private final ZoneId zoneId;
+
+  @Inject
+  public BlobStorage(Credentials credentials, Config config, Environment environment) {
+
+    this.credentials = checkNotNull(credentials);
+    this.container = checkNotNull(config).getString(AZURE_CONTAINER_CONF_PATH);
+    this.accountName = checkNotNull(config).getString(AZURE_STORAGE_ACCT_CONF_PATH);
+    this.zoneId = ZoneId.of(checkNotNull(config).getString(AZURE_REGION_CONF_PATH));
+    this.blobEndpoint = String.format("https://%s.blob.core.windows.net", accountName);
+
+    if (environment.isDev()) {
+      client = new AzuriteClient(config);
+    } else if (environment.isTest()) {
+      client = new NullClient();
+    } else {
+      client = new AzureBlobClient();
+    }
+  }
+
+  @Override
+  public URL getPresignedUrl(String fileName) {
+    String blobUrl = client.getBlobUrl(fileName);
+    String sasToken = client.getSasToken(fileName);
+    String signedUrl = String.format("%s?%s", blobUrl, sasToken);
+    try {
+      return new URL(signedUrl);
+    } catch (java.net.MalformedURLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public Client getClient() {
+    return client;
+  }
+
+  @Override
+  public BlobStorageUploadRequest getSignedUploadRequest(
+      String fileName, String successActionRedirect) {
+    // Azure blob must know the name of a file to generate a SAS for it, so we'll use a UUID
+    // TODO: figure out how to map this to user-uploaded filename
```

Comment:
thought (non-blocking): I *think* so long as we store the original filename in the database associated with the uploaded file it's ok if Azure's version of it is named with a UUID

Reply:
I can open up a PR to do this as soon as this PR is merged. In the meantime I opened https://github.com/seattle-uat/civiform/issues/1768

Reply:
Sounds good!